File: src/main/java/de/thkoeln/inf/gpm/vgb/model/Insurant.java
package de.thkoeln.inf.gpm.vgb.model;


import model.InsurantDao;
import util.DbUtil;

public class Insurant {
    private Long id = null;
    private String name;
    private String firstName;
    private String birthdate;
    private Character sex;
    private Long size;
    private Long weight;
    private Address address;
    private Customer customer;

    public static Insurant createOrUpdate(Insurant insurant) {
        return DbUtil.INSTANCE.runInTransaction(() -> InsurantDao.Companion.save(insurant));
    }

    public static Insurant findById(Long insurantId) {
        return DbUtil.INSTANCE.runInTransaction(() -> InsurantDao.Companion.get(insurantId).toInsurant());
    }

    public static void delete(Long insurantId) {
        DbUtil.INSTANCE.runInTransaction(() -> InsurantDao.Companion.delete(insurantId));
    }

    public Insurant(Long id,
                    String name,
                    String firstName,
                    String birthdate,
                    Character sex,
                    Long size,
                    Long weight,
                    Address address,
                    Customer customer) {
        this.id = id;
        this.name = name;
        this.firstName = firstName;
        this.birthdate = birthdate;
        this.sex = sex;
        this.size = size;
        this.weight = weight;
        this.address = address;
        this.customer = customer;
    }

    public Insurant(String name,
                    String firstName,
                    String birthdate,
                    Character sex,
                    Long size,
                    Long weight,
                    Address address,
                    Customer customer) {
        this.name = name;
        this.firstName = firstName;
        this.birthdate = birthdate;
        this.sex = sex;
        this.size = size;
        this.weight = weight;
        this.address = address;
        this.customer = customer;
    }

    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getBirthdate() {
        return birthdate;
    }

    public void setBirthdate(String birthdate) {
        this.birthdate = birthdate;
    }

    public Character getSex() {
        return sex;
    }

    public void setSex(Character sex) {
        this.sex = sex;
    }

    public Long getSize() {
        return size;
    }

    public void setSize(Long size) {
        this.size = size;
    }

    public Long getWeight() {
        return weight;
    }

    public void setWeight(Long weight) {
        this.weight = weight;
    }

    public Address getAddress() {
        return address;
    }

    public void setAddress(Address address) {
        this.address = address;
    }

    public Customer getCustomer() {
        return customer;
    }

    public void setCustomer(Customer customer) {
        this.customer = customer;
    }
}

File: src/main/java/de/thkoeln/inf/gpm/vgb/model/Insurant.kt
package de.thkoeln.inf.gpm.vgb.model

import org.jetbrains.exposed.dao.IntEntity
import org.jetbrains.exposed.dao.IntEntityClass
import org.jetbrains.exposed.dao.id.EntityID
import org.jetbrains.exposed.dao.id.IntIdTable
import org.jetbrains.exposed.sql.transactions.transaction

class Insurant(
        val id: Int?,
        val name: String,
        val firstName: String,
        val birthdate: String,
        val sex: Char,
        val size: Long,
        val weight: Long,
        val address: Address,
        val customer: Customer
)


class InsurantDao(id: EntityID<Int>) : IntEntity(id) {
    private var name by Insurants.name
    private var firstName by Insurants.firstName
    private var birthdate by Insurants.birthdate
    private var sex by Insurants.sex
    private var size by Insurants.size
    private var weight by Insurants.weight
    private var address by AddressDao referencedOn Insurants.address
    private var customer by CustomerDao referencedOn Insurants.customerId

    companion object : IntEntityClass<InsurantDao>(Insurants) {
        /**
         * Update or create [Address] in database
         */
        fun save(insurant: Insurant): Insurant? = transaction {
            val address = AddressDao.save(insurant.address) ?: return@transaction null
            val addressDao = AddressDao.findById(address.id!!) ?: return@transaction null
            val customerDao = CustomerDao.findById(insurant.customer.id!!) ?: return@transaction null

            val newInsurant = if (insurant.id == null) {
                new { update(insurant, addressDao, customerDao) }
            } else {
                val old = findById(insurant.id)
                old?.update(insurant, addressDao, customerDao)
                findById(insurant.id)
            }

            // set own customer ref, if needed
            if (customerDao.toCustomer().insurantId == null)
                newInsurant?.let { customerDao.setInsurantId(it.id.value) }

            newInsurant?.toInsurant()
        }
    }

    private fun update(insurant: Insurant, addressDao: AddressDao, customerDao: CustomerDao) {
        this.name = insurant.name
        this.firstName = insurant.firstName
        this.birthdate = insurant.birthdate
        this.sex = insurant.sex
        this.size = insurant.size.toInt()
        this.weight = insurant.weight.toInt()
        this.address = addressDao
        this.customer = customerDao

    }

    fun toInsurant() = Insurant(
            id.value,
            name,
            firstName,
            birthdate,
            sex,
            size.toLong(),
            weight.toLong(),
            address.toAddress(),
            customer.toCustomer()
    )
}

object Insurants : IntIdTable() {
    val name = text("Name")
    val firstName = text("Vorname")
    val birthdate = text("Geburtstag")
    val sex = char("Geschlecht")
    val size = integer("Groesse")
    val weight = integer("Gewicht")
    val address = reference("Adresse", Addresses)
    val customerId = reference("Kundennr", Customers)
}